File: Algorithms/src/3721. Longest Balanced Subarray II/Code/solution.js
/**
 * JavaScript (Node / leetcode style)
 * @param {number[]} nums
 * @return {number}
 */
var longestBalanced = function (nums) {
  const n = nums.length;
  const pos = new Map();
  for (let i = 0; i < n; ++i) {
    if (!pos.has(nums[i])) pos.set(nums[i], []);
    pos.get(nums[i]).push(i);
  }

  // Segment tree with mn, mx, lazy
  class SegTree {
    constructor(n) {
      this.n = n;
      this.mn = new Array(4 * n).fill(0);
      this.mx = new Array(4 * n).fill(0);
      this.lazy = new Array(4 * n).fill(0);
    }
    apply(idx, v) {
      this.mn[idx] += v;
      this.mx[idx] += v;
      this.lazy[idx] += v;
    }
    push(idx) {
      const z = this.lazy[idx];
      if (z !== 0) {
        this.apply(idx << 1, z);
        this.apply((idx << 1) | 1, z);
        this.lazy[idx] = 0;
      }
    }
    pull(idx) {
      this.mn[idx] = Math.min(this.mn[idx << 1], this.mn[(idx << 1) | 1]);
      this.mx[idx] = Math.max(this.mx[idx << 1], this.mx[(idx << 1) | 1]);
    }
    addRange(idx, l, r, ql, qr, val) {
      if (ql > qr) return;
      if (ql <= l && r <= qr) {
        this.apply(idx, val);
        return;
      }
      this.push(idx);
      const mid = (l + r) >> 1;
      if (ql <= mid)
        this.addRange(idx << 1, l, mid, ql, Math.min(qr, mid), val);
      if (qr > mid)
        this.addRange(
          (idx << 1) | 1,
          mid + 1,
          r,
          Math.max(ql, mid + 1),
          qr,
          val,
        );
      this.pull(idx);
    }
    add(l, r, v) {
      if (l > r) return;
      this.addRange(1, 0, this.n - 1, l, r, v);
    }
    findRightmostZero(idx, l, r, ql, qr) {
      if (ql > qr || qr < l || ql > r) return -1;
      if (this.mn[idx] > 0 || this.mx[idx] < 0) return -1;
      if (l === r) {
        return this.mn[idx] === 0 ? l : -1;
      }
      this.push(idx);
      const mid = (l + r) >> 1;
      if (qr > mid) {
        const res = this.findRightmostZero(
          (idx << 1) | 1,
          mid + 1,
          r,
          Math.max(ql, mid + 1),
          qr,
        );
        if (res !== -1) return res;
      }
      if (ql <= mid) {
        return this.findRightmostZero(idx << 1, l, mid, ql, Math.min(qr, mid));
      }
      return -1;
    }
    findRightmost(l, r) {
      if (l > r) return -1;
      return this.findRightmostZero(1, 0, this.n - 1, l, r);
    }
  }

  const st = new SegTree(n);
  for (let [val, arr] of pos) {
    const sign = val & 1 ? 1 : -1;
    st.add(arr[0], n - 1, sign);
  }
  const ptr = new Map();
  for (let k of pos.keys()) ptr.set(k, 0);

  let ans = 0;
  for (let l = 0; l < n; ++l) {
    const r = st.findRightmost(l, n - 1);
    if (r !== -1) ans = Math.max(ans, r - l + 1);

    const x = nums[l];
    const pi = ptr.get(x);
    ptr.set(x, pi + 1);
    const arr = pos.get(x);
    const nextPos = pi + 1 < arr.length ? arr[pi + 1] : n;
    const sign = x & 1 ? 1 : -1;
    const L = l,
      R = nextPos - 1;
    if (L <= R) st.add(L, R, -sign);
  }
  return ans;
};

File: Algorithms/src/3721. Longest Balanced Subarray II/Code/solution.go
// Go (clear and efficient)
package main
import (
    "fmt"
)

type SegTree struct {
    n int
    mn, mx, lazy []int
}

func NewSegTree(n int) *SegTree {
    return &SegTree{
        n: n,
        mn: make([]int, 4*n),
        mx: make([]int, 4*n),
        lazy: make([]int, 4*n),
    }
}
func (st *SegTree) apply(idx, v int) {
    st.mn[idx] += v
    st.mx[idx] += v
    st.lazy[idx] += v
}
func (st *SegTree) push(idx int) {
    z := st.lazy[idx]
    if z != 0 {
        st.apply(idx<<1, z)
        st.apply(idx<<1|1, z)
        st.lazy[idx] = 0
    }
}
func (st *SegTree) pull(idx int) {
    if st.mn[idx<<1] < st.mn[idx<<1|1] {
        st.mn[idx] = st.mn[idx<<1]
    } else {
        st.mn[idx] = st.mn[idx<<1|1]
    }
    if st.mx[idx<<1] > st.mx[idx<<1|1] {
        st.mx[idx] = st.mx[idx<<1]
    } else {
        st.mx[idx] = st.mx[idx<<1|1]
    }
}
func (st *SegTree) addRange(idx, l, r, ql, qr, val int) {
    if ql > qr { return }
    if ql <= l && r <= qr {
        st.apply(idx, val); return
    }
    st.push(idx)
    mid := (l + r) >> 1
    if ql <= mid { st.addRange(idx<<1, l, mid, ql, min(qr, mid), val) }
    if qr > mid  { st.addRange(idx<<1|1, mid+1, r, max(ql, mid+1), qr, val) }
    st.pull(idx)
}
func (st *SegTree) Add(l, r, v int) {
    if l > r { return }
    st.addRange(1, 0, st.n-1, l, r, v)
}
func (st *SegTree) findRightmostZero(idx, l, r, ql, qr int) int {
    if ql > qr || qr < l || ql > r { return -1 }
    if st.mn[idx] > 0 || st.mx[idx] < 0 { return -1 }
    if l == r {
        if st.mn[idx] == 0 { return l }
        return -1
    }
    st.push(idx)
    mid := (l + r) >> 1
    if qr > mid {
        res := st.findRightmostZero(idx<<1|1, mid+1, r, max(ql, mid+1), qr)
        if res != -1 { return res }
    }
    if ql <= mid {
        return st.findRightmostZero(idx<<1, l, mid, ql, min(qr, mid))
    }
    return -1
}
func (st *SegTree) FindRightmost(l, r int) int {
    if l > r { return -1 }
    return st.findRightmostZero(1, 0, st.n-1, l, r)
}
func min(a,b int) int { if a<b { return a }; return b }
func max(a,b int) int { if a>b { return a }; return b }

// Example wrapper function (LeetCode expects a function)
func longestBalanced(nums []int) int {
    n := len(nums)
    pos := map[int][]int{}
    for i, v := range nums {
        pos[v] = append(pos[v], i)
    }
    st := NewSegTree(n)
    for v, arr := range pos {
        sign := -1
        if v&1 == 1 { sign = 1 }
        st.Add(arr[0], n-1, sign)
    }
    ptr := map[int]int{}
    for k := range pos { ptr[k] = 0 }
    ans := 0
    for l := 0; l < n; l++ {
        r := st.FindRightmost(l, n-1)
        if r != -1 && r-l+1 > ans { ans = r-l+1 }
        x := nums[l]
        pIndex := ptr[x]; ptr[x] = pIndex+1
        arr := pos[x]
        nextPos := n
        if ptr[x] < len(arr) { nextPos = arr[ptr[x]] }
        sign := -1
        if x&1 == 1 { sign = 1 }
        L, R := l, nextPos-1
        if L <= R { st.Add(L, R, -sign) }
    }
    return ans
}

// main is just for quick local test (remove on leetcode)
func main() {
    fmt.Println(longestBalanced([]int{2,5,4,3})) // expect 4
}
